stdlib/public/core/Misc.swift

```diff
@@ -78,6 +78,24 @@ func _typeName(_ type: Any.Type, qualified: Bool = true) -> String {
     UnsafeBufferPointer(start: stringPtr, count: count)).0
 }
 
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+@_silgen_name("swift_getMangledTypeName")
+public func _getMangledTypeName(_ type: Any.Type)
+  -> (UnsafePointer<UInt8>, Int)
+
+/// Returns the mangled name for a given type.
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+public // SPI
+func _mangledTypeName<T>(_ type: T.Type) -> String? {
+  let (stringPtr, count) = _getMangledTypeName(type)
+  guard count > 0 else {
+    return nil
+  }
+
+  return String._fromUTF8Repairing(
+      UnsafeBufferPointer(start: stringPtr, count: count)).0
+}
```

Comment:
What do we want to do here if the input contains invalid UTF-8? It seems like our mangling should only produce ASCII, so you can use `String._fromASCIIValidating` which returns nil of non-ASCII, and trap in that scenario.

Reply:
CC @eeckstein @DougGregor @jckarter though to verify the ASCII-ness of mangled names for types with non-ASCII names

Reply:
Yes, that's the case. Non-ASCII characters in identifiers are encoded to ASCII with the Punycode algorithm.

Reply:
Punycode encoding is not used in all contexts; we also support disabling it for runtime mangled names, since we aren't constrained by linker limitations there. The mangled string should always be valid UTF-8.

Reply:
So... sounds like nothing to be done here then?

Anything else I can help to get this merged? 

I would be happy to backport as well if possible (?), which branches should I target for "next" 5.1 / 5.2?

Reply:
This LGTM, but I would suggest adding a precondition that the `repairsMade` in the return value was false.

Reply:
Done: https://github.com/apple/swift/pull/30318/commits/34b6e4fe5895a935af438c085f3194c103013932